capa/features/extractors/binexport2/__init__.py

```diff
@@ -0,0 +1,309 @@
+            )
+            self.data_reference_index_by_target_address[data_reference.address].append(data_reference_index)
+
+    def _index_string_references(self):
+        for string_reference_index, string_reference in enumerate(self.be2.string_reference):
+            self.string_reference_index_by_source_instruction_index[string_reference.instruction_index].append(
+                string_reference_index
+            )
+
+    @staticmethod
+    def instruction_indices(basic_block: BinExport2.BasicBlock) -> Iterator[int]:
+        """
+        For a given basic block, enumerate the instruction indices.
+        """
+        for index_range in basic_block.instruction_index:
+            if not index_range.HasField("end_index"):
+                yield index_range.begin_index
+                continue
+            else:
+                yield from range(index_range.begin_index, index_range.end_index)
+
+    def basic_block_instructions(
+        self, basic_block: BinExport2.BasicBlock
+    ) -> Iterator[Tuple[int, BinExport2.Instruction, int]]:
+        """
+        For a given basic block, enumerate the instruction indices,
+        the instruction instances, and their addresses.
+        """
+        instruction_address = 0
+        for instruction_index in self.instruction_indices(basic_block):
+            instruction = self.be2.instruction[instruction_index]
+            if instruction.HasField("address"):
+                instruction_address = instruction.address
+
+            yield instruction_index, instruction, instruction_address
+
+            assert instruction.HasField("raw_bytes")
+            instruction_address += len(instruction.raw_bytes)
+
+    def get_function_name_by_vertex(self, vertex_index: int) -> str:
+        vertex = self.be2.call_graph.vertex[vertex_index]
+        name = f"sub_{vertex.address:x}"
+        if vertex.HasField("mangled_name"):
+            name = vertex.mangled_name
+
+        if vertex.HasField("demangled_name"):
+            name = vertex.demangled_name
+
+        if vertex.HasField("library_index"):
+            library = self.be2.library[vertex.library_index]
+            if library.HasField("name"):
+                name = f"{library.name}!{name}"
+
+        return name
+
+    def get_function_name_by_address(self, address: int) -> str:
+        if address not in self.vertex_index_by_address:
+            return ""
+
+        vertex_index = self.vertex_index_by_address[address]
+        return self.get_function_name_by_vertex(vertex_index)
+
+
+class BinExport2Analysis:
+    def __init__(self, be2: BinExport2, idx: BinExport2Index, buf: bytes):
+        self.be2 = be2
+        self.idx = idx
+        self.buf = buf
+
+        # from virtual address to call graph vertex representing the import
+        self.thunks: Dict[int, int] = {}
+        self.base_address: int = 0
+
+        self._find_got_thunks()
+        self._find_base_address()
+
+    def _find_got_thunks(self):
```

Comment:
@williballenthin can you give some context for why this is needed versus checking for BinExport's vertex type [`Context.Vertex.Type.THUNK`](https://github.com/google/binexport/blob/39f6445c232bb5caf5c4a2a996de91dfa20c48e8/java/src/main/proto/binexport2.proto#L87-L103)?

Reply:
I *think* this is because the exporter wasn't providing that flag, or maybe not recognizing the call to the thunk target. Sorry, I should definitely have included more comments so the context was clear. 

Ultimately, it would be much better for this logic to live in the exporter and use the flag you mentioned. So if that's feasible, let's prefer that path. I'm not at all a fan of maintaining these assembly pattern matches; they were just to demonstrate these things can be done.

Reply:
Totally agree on not matching assembly patterns unless absolutely necessary. Otherwise we lose one of the big benefits of supporting BinExport files.